amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseCleanupManager.java

```diff
@@ -0,0 +1,299 @@
+        for (String parentShard : lease.parentShardIds()) {
+            final Lease parentLease = leaseCoordinator.leaseRefresher().getLease(ShardInfo.getLeaseKey(shardInfo, parentShard));
+
+            if (parentLease != null) {
+                log.warn("Lease {} has a parent lease {} which is still present in the lease table, skipping deletion " +
+                        "for this lease.", lease, parentLease);
+                return false;
+            }
+        }
+        return true;
+    }
+
+    // We should only be deleting the current shard's lease if
+    // 1. All of its children are currently being processed, i.e their checkpoint is not TRIM_HORIZON or AT_TIMESTAMP.
+    // 2. Its parent shard lease(s) have already been deleted.
+    private boolean cleanupLeaseForCompletedShard(Lease lease, ShardInfo shardInfo, Set<String> childShardKeys)
+            throws DependencyException, ProvisionedThroughputException, InvalidStateException, IllegalStateException {
+        final Set<String> processedChildShardLeases = new HashSet<>();
+
+        for (String childShardKey : childShardKeys) {
+            final Lease childShardLease = Optional.ofNullable(leaseCoordinator.leaseRefresher().getLease(childShardKey)).orElseThrow(
+                    () -> new IllegalStateException("Child lease " + childShardKey + " for completed shard not found in " +
+                            "lease table - not cleaning up lease " + lease));
+
+            if (!childShardLease.checkpoint().equals(ExtendedSequenceNumber.TRIM_HORIZON)
+                    && !childShardLease.checkpoint().equals(ExtendedSequenceNumber.AT_TIMESTAMP)) {
+                processedChildShardLeases.add(childShardLease.leaseKey());
+            }
+        }
+
+        if (!allParentShardLeasesDeleted(lease, shardInfo) || !Objects.equals(childShardKeys, processedChildShardLeases)) {
+            return false;
+        }
+
+        log.info("Deleting lease {} as it has been completely processed and processing of child shard(s) has begun.",
+                lease);
+        leaseCoordinator.leaseRefresher().deleteLease(lease);
+
+        return true;
+    }
+
+    private void updateLeaseWithChildShards(LeasePendingDeletion leasePendingDeletion, Set<String> childShardKeys)
+            throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+        final ShardInfo shardInfo = leasePendingDeletion.shardInfo();
+        final Lease updatedLease = leasePendingDeletion.lease().copy();
+        updatedLease.childShardIds(childShardKeys);
+
+        final boolean updated = leaseCoordinator.updateLease(updatedLease, UUID.fromString(shardInfo.concurrencyToken()),
+                LEASE_CLEANUP_MANAGER, ShardInfo.getLeaseKey(leasePendingDeletion.shardInfo()));
+        if (!updated) {
+            throw new InvalidStateException("Failed to update parent lease with child shard information for shard " + shardInfo.shardId());
+        }
```

Comment:
We can use updateLeaseWithMetaData here to directly update this field without any conditional checks.

Reply:
Changed